### PR TITLE
[sema] Add the ability for VarDecls to have parent VarDecls and wire up the VarDecl linked list

### PR DESCRIPTION
**Pull request by gottesmm**
This will enable @akyrtzi and @nathawes to change refactoring/edit-all in scope to use this functionality so that when I add CaseBodyVarDecls refactoring doesn't break and we still have the improved IDE experience.

<rdar://problem/47467128>

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test compiler performance

**Review by nathawes — APPROVED** [review]
Looks good to me! I only noticed a couple of really trivial things going through, so feel free to leave them for later.

File: lib/AST/Decl.cpp [hunk 1]
```diff
@@ -4874,14 +4874,74 @@ SourceRange VarDecl::getTypeSourceRangeForDiagnostics() const {
   return SourceRange();
 }
 
-static bool isVarInPattern(const VarDecl *VD, Pattern *P) {
+static bool isVarInPattern(const VarDecl *vd, Pattern *p) {
   bool foundIt = false;
-  P->forEachVariable([&](VarDecl *FoundFD) {
-    foundIt |= FoundFD == VD;
-  });
+  p->forEachVariable([&](VarDecl *foundFD) { foundIt |= foundFD == vd; });
```
[the remark this answers is not in the document]
Comment (nathawes) [reply lib/AST/Decl.cpp]: ...oh right, it doesn't exist in C++ :-)

**Comment by gottesmm** [comment]
@swift-ci smoke test and merge

**Comment by gottesmm** [comment]
@swift-ci smoke test and merge

**Comment by gottesmm** [comment]
@swift-ci smoke test and merge

**Comment by gottesmm** [comment]
@swift-ci smoke test and merge

**Pull request merged**